src/coord/src/coord.rs

```diff
@@ -3277,46 +3275,65 @@ impl Coordinator {
             session.add_transaction_ops(TransactionOps::Tail)?;
         }
 
-        // Determine the frontier of updates to tail *from*.
-        // Updates greater or equal to this frontier will be produced.
-        let frontier = if let Some(ts) = ts {
-            // If a timestamp was explicitly requested, use that.
-            Antichain::from_elem(
-                self.determine_timestamp(&[source_id], PeekWhen::AtTimestamp(ts))?
-                    .0,
-            )
-        } else {
-            self.determine_frontier(source_id)
+        let make_sink_desc = |coord: &mut Coordinator, from, from_desc, uses| {
+            // Determine the frontier of updates to tail *from*.
+            // Updates greater or equal to this frontier will be produced.
+            let frontier = if let Some(ts) = ts {
+                // If a timestamp was explicitly requested, use that.
+                let ts = coord.determine_timestamp(uses, PeekWhen::AtTimestamp(ts))?;
+                Antichain::from_elem(ts.0)
+            } else {
+                coord.determine_frontier(uses)
+            };
+
+            Ok::<_, CoordError>(SinkDesc {
+                from,
+                from_desc,
+                connector: SinkConnector::Tail(TailSinkConnector::default()),
+                envelope: None,
+                as_of: SinkAsOf {
+                    frontier,
+                    strict: !with_snapshot,
+                },
+            })
         };
-        let sink_name = format!(
-            "tail-source-{}",
-            self.catalog
-                .for_session(session)
-                .humanize_id(source_id)
-                .expect("Source id is known to exist in catalog")
-        );
-        let sink_id = self.catalog.allocate_id()?;
-        session.add_drop_sink(sink_id);
+
+        let dataflow = match from {
+            TailFrom::Id(from_id) => {
+                let from = self.catalog.get_by_id(&from_id);
+                let from_desc = from.desc().unwrap().clone();
+                let sink_id = self.catalog.allocate_id()?;
+                let sink_desc = make_sink_desc(self, from_id, from_desc, &[from_id])?;
+                let sink_name = format!("tail-{}", sink_id);
+                self.dataflow_builder()
+                    .build_sink_dataflow(sink_name, sink_id, sink_desc)?
+            }
+            TailFrom::Query {
+                expr,
+                desc,
+                depends_on,
+            } => {
+                let id = self.allocate_transient_id()?;
+                let expr = self.prep_relation_expr(expr, ExprPrepStyle::Static)?;
+                let desc = RelationDesc::new(expr.typ(), desc.iter_names());
```

Comment:
Does the typ of the expr change? Why do we need a new desc?

Reply:
Yeah it can change post optimization (e.g. keys can get determined and I think nullability info can get refined). There's a comment in `generate_view_ops` or somesuch that you could copy, I think.

Reply:
Ok. I added an assert here and the keys and nullability did indeed change. Didn't know if that mattered. Thankso

Reply:
I'm not totally sure that it does for TAIL, but I copied the thing from `generate_view_ops` to be paranoid.